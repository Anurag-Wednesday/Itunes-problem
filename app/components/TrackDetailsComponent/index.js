import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { Carousel } from 'antd';
import { isEmpty } from 'lodash';
import styled from 'styled-components';
import If from '@components/If';
import T from '@components/T';

const Details = styled.div`
  && {
    display: flex;
    flex-direction: column;
    align-items: center;
    background-color: #231f20;
    color: white;
    margin-left: 7rem;
  }
`;

const TrackTitle = styled.div`
  && {
    text-align: center;
    margin-top: 3rem;
    color: white;
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
  }
`;

const ArtistName = styled.div`
  && {
    text-align: center;
    color: red;
  }
`;

const StyledCarousel = styled(Carousel)`
  && {
    width: 20rem;
    height: 25rem;
    margin-left: 20%;
  }
`;

const TrackImage = styled.img`
  && {
    border-radius: 0.5rem;
    height: 20rem;
    width: 20rem;
  }
`;

const AudioPlayer = styled.audio`
  && {
    width: 20rem;
    height: 6rem;
    color: black;
    margin: 1rem;
  }
`;

const TrackDetailsCard = styled.div`
  && {
    display: flex;
    flex-direction: row;
  }
`;

export function TrackDetailsComponent({
  trackName,
  previewUrl,
  artistName,
  artworkUrl100,
  collectionName,
  primaryGenreName,
  trackViewUrl
}) {
  return (
    <TrackDetailsCard data-testid="track-details-component">
      <If condition={!isEmpty(artworkUrl100)} otherwise={<T data-testid="image_unavailable" id="image_unavailable" />}>
        <StyledCarousel autoplay dotPosition="bottom">
          <div>
            <TrackImage alt={trackName} src={artworkUrl100} />
          </div>
          <div>
            <TrackImage alt={trackName} src={artworkUrl100} />
          </div>
          <div>
            <TrackImage alt={trackName} src={artworkUrl100} />
          </div>
        </StyledCarousel>
      </If>
      <Details>
        <TrackTitle>
          <If condition={!isEmpty(trackName)} otherwise={<T data-testid="no-track-name" id="no_track_name" />}>
            <T type="heading" id="track_details_name" data-testid="track_title" values={{ Track: trackName }} />
          </If>
        </TrackTitle>
        <ArtistName>
          <If condition={!isEmpty(artistName)} otherwise={<T id="artist_data_unavailable" />}>
            <T type="subheading" data-testid="artist_name" id="artist_name" values={{ artist: artistName }} />
          </If>
        </ArtistName>
        <If
          condition={!isEmpty(collectionName)}
          otherwise={<T data-testid="artist_data_unavailable" id="artist_data_unavailable" />}
        >
          <T type="standard" id="track_details_collectionName" values={{ Collection: collectionName }} />
        </If>
        <If condition={!isEmpty(primaryGenreName)} otherwise={<T id="artist_data_unavailable" />}>
          <T type="standard" id="track_details_genre" values={{ Genre: primaryGenreName }} />
        </If>
        <If condition={!isEmpty(previewUrl)} otherwise={<T id="preview_unavailable" />}>
          <AudioPlayer data-testid={trackName} controls>
            <source src={previewUrl} />
          </AudioPlayer>
        </If>
        <If condition={!isEmpty(trackViewUrl)} otherwise={<T id="artist_data_unavailable" />}>
          <a href={trackViewUrl} target="_blank" rel="noreferrer">
            {<T id="track_details_URL" />}
          </a>
        </If>
      </Details>
    </TrackDetailsCard>
  );
}

TrackDetailsComponent.propTypes = {
  artistName: PropTypes.string,
  collectionName: PropTypes.string,
  trackName: PropTypes.string,
  trackViewUrl: PropTypes.string,
  previewUrl: PropTypes.string,
  artworkUrl100: PropTypes.string,
  primaryGenreName: PropTypes.string
};

export default memo(TrackDetailsComponent);
