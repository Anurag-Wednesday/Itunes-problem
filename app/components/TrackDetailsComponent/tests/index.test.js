import React from 'react';
// import { fireEvent } from '@testing-library/dom'
import { renderWithIntl } from '@utils/testUtils';
import TrackDetailsComponent from '../index';
import { translate } from '@app/components/IntlGlobalProvider/index';

describe('<TrackDetailsComponent />', () => {
  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithIntl(<TrackDetailsComponent />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should contain 1 TrackDetailsComponent component', () => {
    const { getAllByTestId } = renderWithIntl(<TrackDetailsComponent />);
    expect(getAllByTestId('track-details-component').length).toBe(1);
  });

  it('should render the track details inside the card', () => {
    const trackName = 'See you Again';
    const artistName = 'Charlie Puth';
    const { getByTestId } = renderWithIntl(<TrackDetailsComponent trackName={trackName} artistName={artistName} />);
    expect(getByTestId('track_title')).toHaveTextContent(trackName);
    expect(getByTestId('artist_name')).toHaveTextContent(artistName);
  });
  it('should render the details unavailable messages in case any props are unavailable or have falsy values', () => {
    const { getByTestId } = renderWithIntl(<TrackDetailsComponent />);
    const noTrackName = translate('no_track_name');
    const artistDataUnavailable = translate('artist_data_unavailable');
    const imageUnavailable = translate('image_unavailable');

    expect(getByTestId('no-track-name')).toHaveTextContent(noTrackName);
    expect(getByTestId('artist_data_unavailable')).toHaveTextContent(artistDataUnavailable);
    expect(getByTestId('image_unavailable')).toHaveTextContent(imageUnavailable);
  });
  it('should render the Carousel and have images in it', () => {
    const artWork =
      'https://is2-ssl.mzstatic.com/image/thumb/Music124/v4/c3/14/d9/c314d9bb-87dd-c1cf-8e31-89e372cf7bf8/source/100x100bb.jpg';
    const { baseElement, getAllByTestId } = renderWithIntl(<TrackDetailsComponent artworkUrl100={artWork} />);
    expect(baseElement.getElementsByClassName('ant-carousel').length).toBe(1);
    expect(getAllByTestId('carousel-image').length).toEqual(7);
  });
});
