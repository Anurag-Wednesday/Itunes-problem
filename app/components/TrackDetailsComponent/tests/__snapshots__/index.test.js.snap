// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackDetailsComponent /> should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="TrackDetailsComponent__TrackDetailsCard-sc-1q5q059-7 ijfgSK"
      data-testid="track-details-component"
    >
      <p
        class="T__StyledText-gjlic1-0 iVJYgn"
        data-testid="image_unavailable"
      >
        Image Unavailable
      </p>
      <div
        class="TrackDetailsComponent__Details-sc-1q5q059-0 ecyiHy"
      >
        <div
          class="TrackDetailsComponent__TrackTitle-sc-1q5q059-2 aNhbq"
        >
          <p
            class="T__StyledText-gjlic1-0 iVJYgn"
            data-testid="no-track-name"
          >
            No track name available
          </p>
        </div>
        <div
          class="TrackDetailsComponent__ArtistName-sc-1q5q059-3 jtjBVl"
        >
          <p
            class="T__StyledText-gjlic1-0 iVJYgn"
            data-testid="t"
          >
            No artist data available
          </p>
        </div>
        <p
          class="T__StyledText-gjlic1-0 iVJYgn"
          data-testid="artist_data_unavailable"
        >
          No artist data available
        </p>
        <p
          class="T__StyledText-gjlic1-0 iVJYgn"
          data-testid="t"
        >
          No artist data available
        </p>
        <p
          class="T__StyledText-gjlic1-0 iVJYgn"
          data-testid="t"
        >
          No preview available
        </p>
        <p
          class="T__StyledText-gjlic1-0 iVJYgn"
          data-testid="t"
        >
          No artist data available
        </p>
      </div>
    </div>
  </div>
</body>
`;
